proposals/push-constants.md

```diff
@@ -0,0 +1,66 @@
+# Subgroups
+
+Status: **Draft**
+
+Last modified: 2024-04-30
+
+Issue: #75
+
+# Requirements
+
+No special requirements.
+
+# WGSL
+
+
+## Address Spaces
+
+| Address space | Sharing among invocations | Default access mode | Notes |
+| --- | --- | --- | --- |
+| `push_constant` | Invocations in [shader stage](https://www.w3.org/TR/WGSL/#shader-stages) | [read](https://www.w3.org/TR/WGSL/#access-read) | For [uniform buffer](https://www.w3.org/TR/WGSL/#uniform-buffer) variables exclude [array types](https://www.w3.org/TR/WGSL/#array-types) variable or [structure types](https://www.w3.org/TR/WGSL/#struct-types) variable contains [array types](https://www.w3.org/TR/WGSL/#array-types) attributes |
+
+
+## Variable and Value Declarations
+
+| Declaration | Mutability | Scope | Effective-value-type | Initializer Support | Initializer Expression | Part of Resource Interface |
+| --- | --- | --- | --- | --- | --- | --- |
+| `var<push_constant>` | Immutable | [Module](https://www.w3.org/TR/WGSL/#module-scope) | [Concrete](https://www.w3.org/TR/WGSL/#type-concrete) [constructible](https://www.w3.org/TR/WGSL/#constructible) [host-shareable](https://www.w3.org/TR/WGSL/#host-shareable) excludes [array types](https://www.w3.org/TR/WGSL/#array-types) and [structure types](https://www.w3.org/TR/WGSL/#struct-types) contains array members | Disallowed	| | Yes. [uniform buffer](https://www.w3.org/TR/WGSL/#uniform-buffer) |
+
+NOTE: Each [entry point](https://www.w3.org/TR/WGSL/#entry-point) can statically use at most one push constant variable.
```

Comment:
Why, exactly?

Reply:
To make implementation a simpler. So we don't need to introduce `@offset` in wgsl to describe multiple push constant variable.

Reply:
I personally would prefer push constants to be like bindings, and not tied to a specific entry point. I tend to keep my bindings/push constants in a separate file I can reuse across passes [like so](https://github.com/bevyengine/bevy/blob/307c694899e88542056fe23458bea672cb288bc8/crates/bevy_pbr/src/meshlet/meshlet_bindings.wgsl#L60).

HLSL seems to have requirements around having push constants be structs anyways https://github.com/gfx-rs/wgpu/issues/5571

Reply:
Was my strawman in https://github.com/gpuweb/gpuweb/issues/75#issuecomment-1916192333 considered at all? If we assume 4-byte elements and alignment, I think this could become a lot easier.

Reply:
@magcius in that proposal, where are `@parameter` variable declared? At the global scope of as function arguments? It seems very similar to specifying an offset, except that it stays in "words" instead of bytes. It'd be nice to support smaller data types in push constants eventually so IMHO it's still be better to act on bytes.

Reply:
>If we want to enforce a single push constant interface we should consider an alternate design of making it a parameter to the entry point (identified via an attribute). If you did consider it, what did you see as the draw backs?

I agree it is an optional design. Personally, I'm slightly prefer current address space mode proposal because:

- It defines a global scope variable (like uniform), so that the helper functions could use it directly. ( instead of accepting attributes as parameter, if we pass push constant variable as entry point function parameter)
- From API aspect, `pushConstantBytes` is related with `PipelineLayout` rather than `ShaderModule`. Take it as a special binding resource is natural.